workflow/common/util.go

```diff
@@ -291,9 +315,54 @@ func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams,
 	}
 	newTmpl.Inputs.Artifacts = newInputArtifacts
 
+	//copy passthrough parameters from Inputs to Arguments
+	if newTmpl.Steps != nil {
```

Comment:
To implement #1525, I'm thinking we should not modify templates directly. Instead, we can modify `Arguments` to pass to `executeTemplate` and `validateTemplate` in validation. What do you think?

e.g.
```diff
diff --git a/pkg/apis/workflow/v1alpha1/workflow_types.go b/pkg/apis/workflow/v1alpha1/workflow_types.go
index c435d5a..bc7a436 100644
--- a/pkg/apis/workflow/v1alpha1/workflow_types.go
+++ b/pkg/apis/workflow/v1alpha1/workflow_types.go
@@ -7,6 +7,7 @@ import (
        "regexp"
        "strings"

+       "github.com/argoproj/argo/errors"
        apiv1 "k8s.io/api/core/v1"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
@@ -1116,6 +1117,29 @@ func (in *Inputs) GetParameterByName(name string) *Parameter {
        return nil
 }

+// GetPassthroughParameters returns parameter argument list where items matching PassthroughRegexp are added from input parameters
+func (in *Inputs) GetPassthroughParameters(argumentParameters []Parameter) ([]Parameter, error) {
+       updatedArgumentParameters := make([]Parameter, 0)
+
+       for _, parameter := range argumentParameters {
+               if parameter.PassthroughRegexp != "" {
+
+                       inputParams, err := in.GetParametersByRegexp(parameter.PassthroughRegexp)
+                       if err != nil {
+                               return nil, errors.Errorf(errors.CodeBadRequest, "regexp %s could not be processed", parameter.PassthroughRegexp)
+                       }
+
+                       for _, inputParam := range inputParams {
+                               newParameter := Parameter{Name: inputParam.Name, Value: inputParam.Value}
+                               updatedArgumentParameters = append(updatedArgumentParameters, newParameter)
+                       }
+               } else {
+                       updatedArgumentParameters = append(updatedArgumentParameters, parameter)
+               }
+       }
+       return updatedArgumentParameters, nil
+}
+
 // HasInputs returns whether or not there are any inputs
 func (in *Inputs) HasInputs() bool {
        if len(in.Artifacts) > 0 {
diff --git a/workflow/common/util.go b/workflow/common/util.go
index 40dd54f..28785ca 100644
--- a/workflow/common/util.go
+++ b/workflow/common/util.go
@@ -311,54 +311,9 @@ func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams,
        }
        newTmpl.Inputs.Artifacts = newInputArtifacts

-       //copy passthrough parameters from Inputs to Arguments
-       if newTmpl.Steps != nil {
-               for i, stepList := range newTmpl.Steps {
-                       for j, step := range stepList {
-                               updatedParameters, err := handlePassthroughParameters(newTmpl.Inputs, step.Arguments.Parameters)
-                               if err != nil {
-                                       return nil, err
-                               }
-                               newTmpl.Steps[i][j].Arguments.Parameters = updatedParameters
-                       }
-               }
-       }
-       if newTmpl.DAG != nil {
-               for i, task := range newTmpl.DAG.Tasks {
-                       updatedParameters, err := handlePassthroughParameters(tmpl.Inputs, task.Arguments.Parameters)
-                       if err != nil {
-                               return nil, err
-                       }
-                       newTmpl.DAG.Tasks[i].Arguments.Parameters = updatedParameters
-               }
-       }
-
        return substituteParams(newTmpl, globalParams, localParams)
 }

-// handlePassthroughParameters returns parameter argument list where items matching PassthroughRegexp are added from input parameters
-func handlePassthroughParameters(inputs wfv1.Inputs, argumentParameters []wfv1.Parameter) ([]wfv1.Parameter, error) {
-       updatedArgumentParameters := make([]wfv1.Parameter, 0)
-
-       for _, parameter := range argumentParameters {
-               if parameter.PassthroughRegexp != "" {
-
-                       inputParams, err := inputs.GetParametersByRegexp(parameter.PassthroughRegexp)
-                       if err != nil {
-                               return nil, errors.Errorf(errors.CodeBadRequest, "regexp %s could not be processed", parameter.PassthroughRegexp)
-                       }
-
-                       for _, inputParam := range inputParams {
-                               newParameter := wfv1.Parameter{Name: inputParam.Name, Value: inputParam.Value}
-                               updatedArgumentParameters = append(updatedArgumentParameters, newParameter)
-                       }
-               } else {
-                       updatedArgumentParameters = append(updatedArgumentParameters, parameter)
-               }
-       }
-       return updatedArgumentParameters, nil
-}
-
 // substituteParams returns a new copy of the template with global, pod, and input parameters substituted
 func substituteParams(tmpl *wfv1.Template, globalParams, localParams map[string]string) (*wfv1.Template, error) {
        tmplBytes, err := json.Marshal(tmpl)
diff --git a/workflow/controller/dag.go b/workflow/controller/dag.go
index 744238a..b684fc4 100644
--- a/workflow/controller/dag.go
+++ b/workflow/controller/dag.go
@@ -404,8 +404,16 @@ func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
                        }
                }

+               args := t.Arguments.DeepCopy()
+               updatedParameters, err := dagCtx.tmpl.Inputs.GetPassthroughParameters(args.Parameters)
+               if err != nil {
+                       woc.initializeNode(nodeName, wfv1.NodeTypeSkipped, task, dagCtx.boundaryID, wfv1.NodeError, err.Error())
+                       return
+               }
+               args.Parameters = updatedParameters
+
                // Finally execute the template
-               _, _ = woc.executeTemplate(taskNodeName, &t, dagCtx.tmplCtx, t.Arguments, dagCtx.boundaryID)
+               _, _ = woc.executeTemplate(taskNodeName, &t, dagCtx.tmplCtx, args, dagCtx.boundaryID)
        }

        if taskGroupNode != nil {
diff --git a/workflow/controller/operator.go b/workflow/controller/operator.go
index 2e51d52..177f56d 100644
--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -205,7 +205,7 @@ func (woc *wfOperationCtx) operate() {

        var workflowStatus wfv1.NodePhase
        var workflowMessage string
-       node, err := woc.executeTemplate(woc.wf.ObjectMeta.Name, &wfv1.Template{Template: woc.wf.Spec.Entrypoint}, woc.tmplCtx, woc.wf.Spec.Arguments, "")
+       node, err := woc.executeTemplate(woc.wf.ObjectMeta.Name, &wfv1.Template{Template: woc.wf.Spec.Entrypoint}, woc.tmplCtx, &woc.wf.Spec.Arguments, "")
        if err != nil {
                // the error are handled in the callee so just log it.
                woc.log.Errorf("%s error in entry template execution: %+v", woc.wf.Name, err)
@@ -233,7 +233,7 @@ func (woc *wfOperationCtx) operate() {
                }
                woc.log.Infof("Running OnExit handler: %s", woc.wf.Spec.OnExit)
                onExitNodeName := woc.wf.ObjectMeta.Name + ".onExit"
-               onExitNode, err = woc.executeTemplate(onExitNodeName, &wfv1.Template{Template: woc.wf.Spec.OnExit}, woc.tmplCtx, woc.wf.Spec.Arguments, "")
+               onExitNode, err = woc.executeTemplate(onExitNodeName, &wfv1.Template{Template: woc.wf.Spec.OnExit}, woc.tmplCtx, &woc.wf.Spec.Arguments, "")
                if err != nil {
                        // the error are handled in the callee so just log it.
                        woc.log.Errorf("%s error in exit template execution: %+v", woc.wf.Name, err)
@@ -1067,7 +1067,7 @@ func (woc *wfOperationCtx) getLastChildNode(node *wfv1.NodeStatus) (*wfv1.NodeSt
 // for the created node (if created). Nodes may not be created if parallelism or deadline exceeded.
 // nodeName is the name to be used as the name of the node, and boundaryID indicates which template
 // boundary this node belongs to.
-func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.TemplateHolder, tmplCtx *templateresolution.Context, args wfv1.Arguments, boundaryID string) (*wfv1.NodeStatus, error) {
+func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.TemplateHolder, tmplCtx *templateresolution.Context, args *wfv1.Arguments, boundaryID string) (*wfv1.NodeStatus, error) {
        woc.log.Debugf("Evaluating node %s: template: %s, boundaryID: %s", nodeName, common.GetTemplateHolderString(orgTmpl), boundaryID)

        node := woc.getNodeByName(nodeName)
@@ -1093,7 +1093,7 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
                localParams[common.LocalVarPodName] = woc.wf.NodeID(nodeName)
        }
        // Do not overwrite tmpl here because the old value is used in the error case.
-       newTmpl, err := common.ProcessArgs(tmpl, &args, woc.globalParams, localParams, false)
+       newTmpl, err := common.ProcessArgs(tmpl, args, woc.globalParams, localParams, false)
        if err != nil {
                return woc.initializeNode(nodeName, wfv1.NodeTypeSkipped, orgTmpl, boundaryID, wfv1.NodeError, err.Error()), err
        }
diff --git a/workflow/controller/steps.go b/workflow/controller/steps.go
index 1c16ff2..2de2e1e 100644
--- a/workflow/controller/steps.go
+++ b/workflow/controller/steps.go
@@ -203,7 +203,16 @@ func (woc *wfOperationCtx) executeStepGroup(stepGroup []wfv1.WorkflowStep, sgNod
                        }
                        continue
                }
-               childNode, err := woc.executeTemplate(childNodeName, &step, stepsCtx.tmplCtx, step.Arguments, stepsCtx.boundaryID)
+
+               args := step.Arguments.DeepCopy()
+               updatedParameters, err := stepsCtx.scope.tmpl.Inputs.GetPassthroughParameters(args.Parameters)
+               if err != nil {
+                       return woc.markNodeError(sgNodeName, err)
+               }
+               args.Parameters = updatedParameters
+
+               childNode, err := woc.executeTemplate(childNodeName, &step, stepsCtx.tmplCtx, args, stepsCtx.boundaryID)
                if err != nil {
                        switch err {
                        case ErrDeadlineExceeded:
diff --git a/workflow/templateresolution/context_test.go b/workflow/templateresolution/context_test.go
index a311561..e78d9be 100644
--- a/workflow/templateresolution/context_test.go
+++ b/workflow/templateresolution/context_test.go
@@ -76,14 +76,14 @@ spec:
     template: whalesay-with-passthrough-arguments-inner
     inputs:
       parameters:
-      - passthroughRegexp: ".*"
+      - regexp: ".*"
     arguments:
       parameters:
       - passthroughRegexp: ".*"
   - name: whalesay-with-passthrough-arguments-inner
     inputs:
       parameters:
-      - passthroughRegexp: ".*"
+      - regexp: ".*"
     container:
       image: docker/whalesay
       command: [cowsay]
diff --git a/workflow/validate/validate.go b/workflow/validate/validate.go
index 8f1da74..ce29e71 100644
--- a/workflow/validate/validate.go
+++ b/workflow/validate/validate.go
@@ -338,9 +338,7 @@ func validateInputs(tmpl *wfv1.Template, extraScope map[string]interface{}) (map
                scope[name] = value
        }
        for _, param := range tmpl.Inputs.Parameters {
-               if param.Regexp != "" {
-                       scope[fmt.Sprintf("inputs.parameters.regexp.%s", param.Regexp)] = true
-               } else {
+               if param.Name != "" {
                        scope[fmt.Sprintf("inputs.parameters.%s", param.Name)] = true
                }
        }
@@ -394,14 +392,6 @@ func resolveAllVariables(scope map[string]interface{}, tmplStr string) error {
        var unresolvedErr error
        _, allowAllItemRefs := scope[anyItemMagicValue] // 'item.*' is a magic placeholder value set by addItemsToScope

-       //inputs.parameters.regexp. stores regexps for input parameters
-       inputRegexps := make([]string, 0)
-       for k := range scope {
-               if strings.HasPrefix(k, "inputs.parameters.regexp.") {
-                       inputRegexps = append(inputRegexps, strings.Replace(k, ".regexp", "", -1))
-               }
-       }
-
        fstTmpl := fasttemplate.New(tmplStr, "{{", "}}")

        fstTmpl.ExecuteFuncString(func(w io.Writer, tag string) (int, error) {
@@ -417,20 +407,7 @@ func resolveAllVariables(scope map[string]interface{}, tmplStr string) error {
                                // NOTE: this is far from foolproof.
                        } else if strings.HasPrefix(tag, common.GlobalVarWorkflowCreationTimestamp) {
                        } else {
-                               //check for regexp matches
-                               foundRegexpMatch := false
-                               for _, regexpToMatch := range inputRegexps {
-                                       match, err := regexp.MatchString(regexpToMatch, tag)
-                                       if err != nil {
-                                               return 0, err
-                                       }
-                                       if match {
-                                               foundRegexpMatch = true
-                                       }
-                               }
-                               if foundRegexpMatch == false {
-                                       unresolvedErr = fmt.Errorf("failed to resolve {{%s}}", tag)
-                               }
+                               unresolvedErr = fmt.Errorf("failed to resolve {{%s}}", tag)
                        }
                }
                return 0, nil
@@ -585,7 +562,15 @@ func (ctx *templateValidationCtx) validateSteps(scope map[string]interface{}, tm
                        if err != nil {
                                return err
                        }
-                       resolvedTmpl, err := ctx.validateTemplateHolder(&step, tmplCtx, &step.Arguments, scope)
+
+                       args := step.Arguments.DeepCopy()
+                       updatedParameters, err := tmpl.Inputs.GetPassthroughParameters(args.Parameters)
+                       if err != nil {
+                               return err
+                       }
+                       args.Parameters = updatedParameters
+
+                       resolvedTmpl, err := ctx.validateTemplateHolder(&step, tmplCtx, args, scope)
                        if err != nil {
                                return errors.Errorf(errors.CodeBadRequest, "templates.%s.steps[%d].%s %s", tmpl.Name, i, step.Name, err.Error())
                        }
@@ -1049,8 +1034,16 @@ func (ctx *templateValidationCtx) validateDAG(scope map[string]interface{}, tmpl
                if err != nil {
                        return err
                }
+
+               args := task.Arguments.DeepCopy()
+               updatedParameters, err := tmpl.Inputs.GetPassthroughParameters(args.Parameters)
+               if err != nil {
+                       return err
+               }
+               args.Parameters = updatedParameters
+
                // Validate the template again with actual arguments.
-               _, err = ctx.validateTemplateHolder(&task, tmplCtx, &task.Arguments, scope)
+               _, err = ctx.validateTemplateHolder(&task, tmplCtx, args, scope)
                if err != nil {
                        return errors.Errorf(errors.CodeBadRequest, "templates.%s.tasks.%s %s", tmpl.Name, task.Name, err.Error())
                }
diff --git a/workflow/validate/validate_test.go b/workflow/validate/validate_test.go
index d7e4010..13722d7 100644
--- a/workflow/validate/validate_test.go
+++ b/workflow/validate/validate_test.go
@@ -294,7 +294,7 @@ func TestPassthrough(t *testing.T) {

        err = validate(badArgsRegex)
        if assert.NotNil(t, err) {
-               assert.Contains(t, err.Error(), "regexp ( could not be processed")
+               assert.Contains(t, err.Error(), "error parsing regexp: missing closing ): `(`")
        }

        err = validate(badInputRegex)
```

Reply:
Maybe, we have to make like above for `ProcessArgs` instead of `executeTemplate` and `validateTemplate`.

Reply:
Good idea, I've just included this, with the same logic in MergeReferredTemplate.